### DISCUSSION TITLE
unable to start the service in services.msc

### DISCUSSION THREAD

**Discussion by p060477** [DNSCrypt/dnscrypt-proxy#2282] [Q&A]
hi, following yr:
https://github.com/dnscrypt/dnscrypt-proxy/wiki/Installation-Windows
all went well
and also step 5:
 install the proxy as a system service
went well:
see my dos:
N:\DESKTOP\PC\Dns\dnscrypt-proxy>dnscrypt-proxy -service install
[2023-01-09 17:01:12] [NOTICE] Installed as a service. Use `-service start` to start

N:\DESKTOP\PC\Dns\dnscrypt-proxy>dnscrypt-proxy -service start
[2023-01-09 17:01:52] [NOTICE] Service started

but if i start services.msc in windows 10 i see the service "DNSCrypt" in my services but is still stopped...and if i try to start i get a pop up error:
error 1067

where is my fault...??
apologize for my very few skills... :(
thxs in adv also for yr patience...
cheers

**Comment by jedisct1** [comment]
Maybe an error in the configuration file?

In [step 4](https://github.com/dnscrypt/dnscrypt-proxy/wiki/Installation-Windows#step-4-tweak-the-configuration-file), did everything work as expected when you typed `start dnscrypt-proxy`?

**Comment by p060477** [comment]
i really do not done any changes in the  config.file...so i do not follow the step 4:Tweak the configuration file...i jumped from step 3 to step 5....
this is a try just done...

N:\DESKTOP\PC\Dns\dnscrypt-proxy>dnscrypt-proxy -resolve quad9.com
Resolving [quad9.com] using 127.0.0.1 port 53

Resolver      : 162.158.196.43

Canonical name: quad9.com.

IPv4 addresses: 216.21.3.77
IPv6 addresses: 2620:0:871:9000::77

Name servers  : ns3.pch.net., anyns.pch.net., ns2.pch.net.
DNSSEC signed : no
Mail servers  : 2 mail servers found

HTTPS alias   : -
HTTPS info    : -

Host info     : -
TXT records   : v=spf1 a mx ip4:216.21.3.0/24 ?all

**Comment by lifenjoiner** [comment]
`dnscrypt-proxy -resolve quad9.com` is answered, means there is a server running correctly.
Did you terminate (step 5) the server that is started by previous step?

**Comment by p060477** [comment]
Hi, first thxs so much for yr kind attention,
having reeboot now in services.msc i'm able to finally start the service, but, even if in its properties is set to :"automatic"
at avery reboot i find it -stopped- and i have to start always it by myself...why?
another question: is there a reason for choosing as primary dns the 127.0.0.1 ?
why the 127.0.0.1? apologize my very few skills and level and so also the level and quality of my questions.. :(
here it is another my try:
N:\DESKTOP\PC\Dns\dnscrypt-proxy>dnscrypt-proxy -resolve google.it
Resolving [google.it] using 127.0.0.1 port 53

Resolver      : 165.232.32.95 (do.dnscrypt.uk.)

Canonical name: google.it.

IPv4 addresses: 142.250.187.195
IPv6 addresses: 2a00:1450:4009:81f::2003

Name servers  : ns1.google.com., ns4.google.com., ns2.google.com., ns3.google.com.
DNSSEC signed : no
Mail servers  : 1 mail servers found

HTTPS alias   : -
HTTPS info    : -

Host info     : -
TXT records   : v=spf1 -all

thxs againg for yr kind patience
cheers

**Comment by lifenjoiner** [comment]
`127.0.0.1`:
You will know why, if you do some searches.
It is the most used IPv4 address for **localhost**, which is a safe server IP for the same machine clients. Of course, you can set it listens to other network interface(s) and set it(one) as your primary DNS, if needed.

Service mode:
It works, in my tests.
So,
1. you'd better describe more details of your environments, to see if any other people has a same issue or could reproduce it;
2. take a look at the event logs according event viewer, to see if there is any meaningful message about the service, and do some research by yourself to make any progress ...

**Comment by p060477** [comment]
hi, first thxs for yr attention, i have previously beg your pardon and apologize for my very few and poor skills...
1) i'm interested in yr opinion, for you the better choice is to set primary dns 127.0.0.1? If  right no need any other words then -yes- and i will do without bothering you again , i promise

2) in event viewer there is an error about dnscrypt-proxy service:
Nome registro: System
Origine:       Service Control Manager
Data:          11/01/2023 16:46:33
ID evento:     7009
Categoria attività:Nessuna
Livello:       Errore
Parole chiave: Classico
Utente:        N/D
Computer:      pc
Descrizione:
Timeout (45000 millisecondi) durante l'attesa della connessione del servizio dnscrypt-proxy.
XML evento:
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
  <System>
    <Provider Name="Service Control Manager" Guid="{555908d1-a6d7-4695-8e1e-26931d2012f4}" EventSourceName="Service Control Manager" />
    <EventID Qualifiers="49152">7009</EventID>
    <Version>0</Version>
    <Level>2</Level>
    <Task>0</Task>
    <Opcode>0</Opcode>
    <Keywords>0x8080000000000000</Keywords>
    <TimeCreated SystemTime="2023-01-11T15:46:33.8147724Z" />
    <EventRecordID>45425</EventRecordID>
    <Correlation />
    <Execution ProcessID="984" ThreadID="1576" />
    <Channel>System</Channel>
    <Computer>pc</Computer>
    <Security />
  </System>
  <EventData>
    <Data Name="param1">45000</Data>
    <Data Name="param2">dnscrypt-proxy</Data>
    <Binary>64006E007300630072007900700074002D00700072006F00780079000000</Binary>
  </EventData>
</Event>

my details are:
pc desktop windows 10 home 64 , with the gpedit opt. due to a reg modify,
vers 22H2 (build S.O 19045.2486
if you need more details pls feel free to ask me, thxs.


i re-pray you to have the patience to help me also if i'm not so skilled....

thxs so much indeed in adv and waiting for yr kind helping reply i wish you all the best
cheers

**Comment by lifenjoiner** [comment]
127.0.0.1:
Yes, in my scenario.
<br>

Service:
```
Descrizione:
Timeout (45000 millisecondi) durante l'attesa della connessione del servizio dnscrypt-proxy.
```
Italian to English:
```
Description:
Timeout (45000 milliseconds) while waiting for dnscrypt-proxy service connection.
```
Next is to find out why it is `timeout`. Uncomment
https://github.com/DNSCrypt/dnscrypt-proxy/blob/5438eed2f43e0ac26cc3c441f680faecbaf4762d/dnscrypt-proxy/example-dnscrypt-proxy.toml#L173
and restart to see if we can get any more useful message.

~~
An alternative try: change the service start type from `Automatic` to `Automatic (Delayed Start)`.

**Comment by p060477** [comment]
hi again , i re apologize but i really re-pray you to be more easy and to think to speak to a child...i'm very sorry but what have i to do in the config file .toml...??....which are the modifications have i to try..??..
here is a my try : 
![Immagine](https://user-images.githubusercontent.com/12773559/212032313-a1772309-355e-4ed7-88a1-237f86139a98.jpg)

but -nothing- change...at re start the service is stopped...the event viewer is the -same- with same error
and i really do not find any dnscrypt-proxy.log file in anyplace of my pc...
could you pls have the gentle kindness to write it here down in a very easy way..??...which are the line-or lines have i to change..??..and in which way...?? thxs so much indeed for yr kind understanding
i really beg yr pardon again. cheers

**Comment by lifenjoiner** [comment]
You did it right.
We are trying to follow the vine to get the melon -- track down the failure by following clues. Of course the same error, as we just turned on logging to a file.
If `dnscrypt-proxy` is ever launched, there should be the `dnscrypt-proxy.log` in the same folder. If there isn't any, the service seems not be launched. Strange, and this direction ends!

There are several reasons on the web for this "Event ID 7009" timeout error, but all uncertain.

You say you did "gpedit opt. due to a reg modify", I don't know if it leads to this ...

Try change the service start type from `Automatic` to `Automatic (Delayed Start)`, to see if it help.

These [discussions](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/ebea5f1a-14fe-4063-bd38-31f8c3396a06/a-timeout-was-reached-30000-milliseconds-while-waiting-for-the-service-service-to-connect?forum=csharpgeneral) may be the closest to the truth, in my opinion. Maybe you can try [increase the timeout period](https://social.technet.microsoft.com/wiki/contents/articles/1464.event-id-7009-basic-service-operations.aspx), but still not sure. Is your PC slow?

No other ideas ...

**Comment by lifenjoiner** [comment]
One more test in CLI:
```
set GODEBUG=inittrace=1
dnscrypt-proxy
```
What is it before line `[NOTICE] dnscrypt-proxy 2.1.2`?

**Comment by p060477** [comment]
Hi and thxs so much again
here it is the  command line try : 
if i start:

N:\DESKTOP\PC\Dns\dnscrypt-proxy>set GODEBUG=inittrace=1

N:\DESKTOP\PC\Dns\dnscrypt-proxy>dnscrypt-proxy:

init github.com/lucas-clemente/quic-go/internal/handshake @253 ms, 0 ms clock, 1584 bytes, 13 allocs
init github.com/lucas-clemente/quic-go @253 ms, 0 ms clock, 64 bytes, 4 allocs
init golang.org/x/net/http2/hpack @253 ms, 0 ms clock, 22296 bytes, 32 allocs
init github.com/marten-seemann/qpack @254 ms, 0 ms clock, 10640 bytes, 68 allocs
init golang.org/x/text/transform @254 ms, 0 ms clock, 80 bytes, 5 allocs
init golang.org/x/text/unicode/bidi @255 ms, 0 ms clock, 272 bytes, 2 allocs
init golang.org/x/text/secure/bidirule @255 ms, 0 ms clock, 16 bytes, 1 allocs
init golang.org/x/text/unicode/norm @256 ms, 0 ms clock, 0 bytes, 0 allocs
init golang.org/x/net/idna @256 ms, 0 ms clock, 0 bytes, 0 allocs
init golang.org/x/net/http/httpguts @258 ms, 0 ms clock, 688 bytes, 2 allocs
init github.com/lucas-clemente/quic-go/http3 @258 ms, 0 ms clock, 64 bytes, 4 allocs
init golang.org/x/net/http2 @258 ms, 0 ms clock, 4336 bytes, 63 allocs
init main @259 ms, 0 ms clock, 832 bytes, 8 allocs

and here there is the log file:

[2023-01-12 20:32:19] [NOTICE] dnscrypt-proxy 2.1.2
[2023-01-12 20:32:19] [NOTICE] Network connectivity detected
[2023-01-12 20:32:19] [NOTICE] Now listening to 127.0.0.1:53 [UDP]
[2023-01-12 20:32:19] [NOTICE] Now listening to 127.0.0.1:53 [TCP]
[2023-01-12 20:32:19] [NOTICE] Source [relays] loaded
[2023-01-12 20:32:19] [NOTICE] Source [public-resolvers] loaded
[2023-01-12 20:32:19] [NOTICE] Firefox workaround initialized
[2023-01-12 20:32:21] [NOTICE] [jp.tiarap.org] OK (DoH) - rtt: 318ms
[2023-01-12 20:32:21] [NOTICE] [doh.ffmuc.net-2] OK (DoH) - rtt: 140ms
[2023-01-12 20:32:22] [NOTICE] [dct-nl1] OK (DNSCrypt) - rtt: 74ms
[2023-01-12 20:32:22] [NOTICE] [dnscrypt.ca-2] OK (DNSCrypt) - rtt: 143ms
[2023-01-12 20:32:22] [NOTICE] [jp.tiar.app] OK (DNSCrypt) - rtt: 286ms
[2023-01-12 20:32:22] [NOTICE] [openinternet] OK (DNSCrypt) - rtt: 212ms
[2023-01-12 20:32:22] [NOTICE] [scaleway-ams] OK (DNSCrypt) - rtt: 56ms
[2023-01-12 20:32:22] [NOTICE] [faelix-ch-ipv4] OK (DNSCrypt) - rtt: 45ms
[2023-01-12 20:32:23] [NOTICE] [sby-limotelu] OK (DNSCrypt) - rtt: 206ms
[2023-01-12 20:32:23] [NOTICE] [quad9-doh-ip4-port443-nofilter-pri] OK (DoH) - rtt: 23ms
[2023-01-12 20:32:23] [NOTICE] [doh-crypto-sx] OK (DoH) - rtt: 74ms
[2023-01-12 20:32:23] [NOTICE] [ibksturm] OK (DNSCrypt) - rtt: 78ms
[2023-01-12 20:32:24] [NOTICE] [altername] OK (DNSCrypt) - rtt: 92ms
[2023-01-12 20:32:24] [NOTICE] [sth-dnscrypt-se] OK (DNSCrypt) - rtt: 86ms
[2023-01-12 20:32:24] [NOTICE] [controld-unfiltered] OK (DoH) - rtt: 43ms

what you think...??

the strange thing is that the service in services.msc is stopped...but if i try in command line
N:\DESKTOP\PC\Dns\dnscrypt-proxy>dnscrypt-proxy -resolve cloudfare.com
Resolving [cloudfare.com] using 127.0.0.1 port 53

Resolver      : 89.38.131.38 (alekberg.net.)

Canonical name: cloudfare.com.

IPv4 addresses: 188.114.97.12, 188.114.96.12
IPv6 addresses: 2a06:98c1:3121::3, 2a06:98c1:3120::3

Name servers  : fish.ns.cloudflare.com., wells.ns.cloudflare.com.
DNSSEC signed : no
Mail servers  : 4 mail servers found

HTTPS alias   : -
HTTPS info    : [alpn]=[h3,h3-29,h2], [ipv4hint]=[104.21.77.216,172.67.211.231], [ipv6hint]=[2606:4700:3031::6815:4dd8,2606:4700:3034::ac43:d3e7]

Host info     : -
TXT records   : -


N:\DESKTOP\PC\Dns\dnscrypt-proxy>

seems to works even if not started...for my is a very mistery...

cause it is -stopped-...:

![Immagine 3](https://user-images.githubusercontent.com/12773559/212176821-bfab8908-4a02-4e82-be31-17ce03c83484.jpg)


i apologize me again and 
i re-thxs you so much again indeed
cheers

**Comment by lifenjoiner** [comment]
There are conflicts I don't understand either.
Check if there is the `dnscrypt-proxy` process running in a task manager. If yes, it may be some bug of the service manager.

**Reply by p060477** [reply]
thxs so much again for yr attention
here are my tries:

1) start pc : 
no service running in services.msc and 
no dnscrypt-proxy in task manager and 
the command line dnscrypt-proxy -resolve quad9.net  fails and 
in the .log file there is the old past activity

2) i run in command line yr suggestion: 
set GODEBUG=inittrace=1
dnscrypt-proxy

3) after running that-> 
in task manager appears the dnscrypt-proxy.exe process
the command dnscrypt-proxy -resolve quad9.net  works 
the .log file now is updated to this new activity
but the service in services.msc is still stopped

what you think about?  yr opinion is very very precious for me
thxs so much again
cheers

**Comment by lifenjoiner** [comment]
I see. Your step 3 works, because of you start `dnscrypt-proxy` in step 2.

So, the service is not started successfully, but timeout.
The `init main @259 ms` shows your PC is a bit slow, while it is `init main @102 ms` with Intel CPU i7-3612QM. [Slow CPU would cause the timeout](https://github.com/grafana/agent/issues/2060).

Solutions:
1. As my previous suggestions, it maybe work, and worth a try.
1.1. [Increase the timeout period](https://social.technet.microsoft.com/wiki/contents/articles/1464.event-id-7009-basic-service-operations.aspx).
1.2. Change the service start type from `Automatic` to `Automatic (Delayed Start)`. In a less load period, it maybe start up quicker.
2. Abort service mode, and use a third party tool to launch `dnscrypt-proxy` in hidden mode.

**Comment by p060477** [comment]
Hi and thxs so much again,
about yr reply:
1)
"your PC is a bit slow"...: 
is there a way to hurry it up...??...and as much easy as for a not very skilled like me..??
i've Intel CPU i5 4590 of 2015
2)
"1.1. [Increase the timeout period](https://social.technet.microsoft.com/wiki/contents/articles/1464.event-id-7009-basic-service-operations.aspx). "..:
which value you will suggest to choose? 
3)
"2.Abort service mode, and use a third party tool to launch dnscrypt-proxy in hidden mode."...:
how to..??
may you have the patience to explain it as you were talking to a little  "not- skilled- at- all"  child..??
4)
and in your scenario is better to have the default line:
"# server_names = ['scaleway-fr', 'google', 'yandex']"
or to choose one specific as,for example : 
"server_names = ['quad9-doh-ip4-port443-filter-pri', 'cloudflare']"
also in this case for me is mainly important your opinion
i'm interested in yr opinion, for you which the better choice
and so which is the one you have choosen?
5)
i'm very interested in realizing if all this work  has a positive result so i check yr istructions here:
https://github.com/dnscrypt/dnscrypt-proxy/wiki/Checking
these are the results of my trying that are really no good at all.. :( 
- stop/pause the proxy: i kill the process, dnscrypt-proxy.exe  in task manager but still able to navigate on the web without issues
- i try also enable query logging: https://github.com/dnscrypt/dnscrypt-proxy/wiki/Logging 
  changing the config file in this way:

**original:**

![Immagine](https://user-images.githubusercontent.com/12773559/212429015-2bc82df5-cdc7-47a9-8cdb-691fdcd6f97f.jpg)



**modified**

![Immagine2](https://user-images.githubusercontent.com/12773559/212429073-9fa7a0e5-5b0f-4ed8-8316-f977bf2d8363.jpg)


but in this way the service stopped in services.msc...and the dnscrypt-proxy.exe process stopped in task manager
and no trace about a query.log file in my pc...
and no trace also in dnscrypt-proxy.log file activity....where is my fault..??.. :( 

the third party's test here:
https://www.dnsleaktest.com/results.html
give always this result, i made -4- extended tests:
<html><body>


IP | Hostname | ISP | Country
-- | -- | -- | --
172.71.113.77 | None | Cloudflare | Milan, Italy


</body>
</html>

but i previously have made also this change in the config .toml file:

server_names = ['quad9-doh-ip4-port443-filter-pri', 'cloudflare']

but is strange that having choosen 2 servers, quad9 and cloudfare the response about the 4 tries are always the same ....
seems that only cloudfare is working...
but  the logfile instead states a different thing...
here it is:

2023-01-13 22:36:48] [NOTICE] dnscrypt-proxy 2.1.2
[2023-01-13 22:36:48] [NOTICE] Network connectivity detected
[2023-01-13 22:36:48] [NOTICE] Now listening to 127.0.0.1:53 [UDP]
[2023-01-13 22:36:48] [NOTICE] Now listening to 127.0.0.1:53 [TCP]
[2023-01-13 22:36:49] [NOTICE] Source [public-resolvers] loaded
[2023-01-13 22:36:51] [NOTICE] Source [relays] loaded
[2023-01-13 22:36:51] [NOTICE] Firefox workaround initialized
[2023-01-13 22:36:51] [NOTICE] [quad9-doh-ip4-port443-filter-pri] OK (DoH) - rtt: 23ms
[2023-01-13 22:36:52] [NOTICE] [cloudflare] OK (DoH) - rtt: 24ms
[2023-01-13 22:36:52] [NOTICE] Sorted latencies:
[2023-01-13 22:36:52] [NOTICE] -    23ms quad9-doh-ip4-port443-filter-pri
[2023-01-13 22:36:52] [NOTICE] -    24ms cloudflare
[2023-01-13 22:36:52] [NOTICE] Server with the lowest initial latency: quad9-doh-ip4-port443-filter-pri (rtt: 23ms)
[2023-01-13 22:36:52] [NOTICE] dnscrypt-proxy is ready - live servers: 2

so to sum up i'm really not sure at all  if my DNS traffic is encrypted and authenticated...what you think about...??



thxs so much indeed and in deep again
also for yr kind and very precious helping and big patience too
cheers
